sw/device/tests/autogen/plic_all_irqs_test.c

```diff
@@ -0,0 +1,489 @@
+    //   break;
+    // case kTopEarlgreyPlicPeripheralSpiDevice:
+    //   PERIPHERAL_ISR(spi_device, spi_device, kTopEarlgreyPlicIrqIdSpiDeviceRxf);
+    //   break;
+    // case kTopEarlgreyPlicPeripheralSpiHost0:
+    //   PERIPHERAL_ISR(spi_host, spi_host0, kTopEarlgreyPlicIrqIdSpiHost0Error);
+    //   break;
+    // case kTopEarlgreyPlicPeripheralSpiHost1:
+    //   PERIPHERAL_ISR(spi_host, spi_host1, kTopEarlgreyPlicIrqIdSpiHost1Error);
+    //   break;
+    // case kTopEarlgreyPlicPeripheralSysrstCtrlAon:
+    //   PERIPHERAL_ISR(sysrst_ctrl, sysrst_ctrl_aon, kTopEarlgreyPlicIrqIdSysrstCtrlAonSysrstCtrl);
+    //   break;
+    case kTopEarlgreyPlicPeripheralUart0:
+      PERIPHERAL_ISR(uart, uart0, kTopEarlgreyPlicIrqIdUart0TxWatermark);
+      break;
+    case kTopEarlgreyPlicPeripheralUart1:
+      PERIPHERAL_ISR(uart, uart1, kTopEarlgreyPlicIrqIdUart1TxWatermark);
+      break;
+    case kTopEarlgreyPlicPeripheralUart2:
+      PERIPHERAL_ISR(uart, uart2, kTopEarlgreyPlicIrqIdUart2TxWatermark);
+      break;
+    case kTopEarlgreyPlicPeripheralUart3:
+      PERIPHERAL_ISR(uart, uart3, kTopEarlgreyPlicIrqIdUart3TxWatermark);
+      break;
+    // case kTopEarlgreyPlicPeripheralUsbdev:
+    //   PERIPHERAL_ISR(usbdev, usbdev, kTopEarlgreyPlicIrqIdUsbdevPktReceived);
+    //   break;
+    default:
+      LOG_FATAL("ISR is not implemented!");
+      test_status_set(kTestStatusFailed);
+  }
+
+#undef PERIPHERAL_ISR
+
+  // Complete the IRQ at PLIC.
+  CHECK(dif_rv_plic_irq_complete(&plic, kHart, plic_irq_id) == kDifRvPlicOk,
+        "dif_rv_plic_irq_complete failed");
+}
+
+/**
+ * Enables all IRQs at PLIC.
+ */
+static void plic_irqs_enable(void) {
+  mmio_region_t plic_base_addr =
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR);
+  CHECK(dif_rv_plic_init((dif_rv_plic_params_t){.base_addr = plic_base_addr},
+                         &plic) == kDifRvPlicOk,
+        "dif_rv_plic_init failed");
+
+  for (top_earlgrey_plic_irq_id_t irq_id = kTopEarlgreyPlicIrqIdNone + 1;
+       irq_id <= kTopEarlgreyPlicIrqIdLast; ++irq_id) {
+    // TODO: pick priority randomly between 0-3.
+    CHECK(dif_rv_plic_irq_set_priority(&plic, irq_id, 0x1) == kDifRvPlicOk,
+          "dif_rv_plic_irq_set_priority failed");
+
+    CHECK(dif_rv_plic_irq_set_enabled(&plic, irq_id, kHart,
+                                      kDifRvPlicToggleEnabled) == kDifRvPlicOk,
+          "dif_rv_plic_irq_set_enabled failed");
+  }
+  CHECK(dif_rv_plic_target_set_threshold(&plic, kHart, 0x0) == kDifRvPlicOk,
+        "dif_rv_plic_target_set_threshold failed");
+}
+
+/**
+ * Initializes the handles to all peripherals.
+ */
+static void peripherals_init(void) {
+#define PERIPHERAL_INIT(_peripheral, _handle, _base_addr)          \
+  do {                                                             \
+    mmio_region_t base_addr = mmio_region_from_addr(_base_addr);   \
+    CHECK(dif_##_peripheral##_init(base_addr, &_handle) == kDifOk, \
```

Comment:
Just a heads up/FYI, this should be the DIF init function prototype for _most_ IPs, however for IPs that are templated, thus have HW instantiation params (*_params_t), there will be an extra parameter in this function prototype. See #8286 for an example with the alert_handler. Just means that this will have some special cases thats all... square peg, round hole...